File: HabitCoachRN/src/services/TimerService.ts
import { SessionEndCondition } from '../models/types';

export interface TimerState {
  isRunning: boolean;
  elapsedSeconds: number;
  reminderCount: number;
  secondsUntilNextBuzz: number;
  isComplete: boolean;
}

export interface TimerConfig {
  intervalSeconds: number;
  varianceSeconds: number;
  endCondition: SessionEndCondition;
  onInterval: () => void;
  onComplete?: () => void;
}

export class TimerService {
  private timer: ReturnType<typeof setInterval> | null = null;
  private nextBuzzAt = 0;
  private config: TimerConfig | null = null;
  private sessionStartDate: Date | null = null;
  private _state: TimerState = {
    isRunning: false,
    elapsedSeconds: 0,
    reminderCount: 0,
    secondsUntilNextBuzz: 0,
    isComplete: false,
  };
  private onStateChange: (state: TimerState) => void;

  constructor(onStateChange: (state: TimerState) => void) {
    this.onStateChange = onStateChange;
  }

  get state(): TimerState {
    return { ...this._state };
  }

  get totalCycles(): number | null {
    if (!this.config) return null;
    const { endCondition, intervalSeconds } = this.config;
    switch (endCondition.type) {
      case 'unlimited':
        return null;
      case 'afterCount':
        return endCondition.count;
      case 'afterDuration':
        return intervalSeconds > 0 ? Math.floor(endCondition.seconds / intervalSeconds) : null;
    }
  }

  start(config: TimerConfig) {
    if (config.intervalSeconds <= 0) return;
    this.config = config;
    this._state = {
      isRunning: true,
      elapsedSeconds: 0,
      reminderCount: 0,
      secondsUntilNextBuzz: 0,
      isComplete: false,
    };
    this.sessionStartDate = new Date();
    this.scheduleNextBuzz();
    this.startTicking();
    this.emit();
  }

  reconfigure(config: TimerConfig) {
    this.config = config;
    this._state.elapsedSeconds = 0;
    this._state.reminderCount = 0;
    this._state.isComplete = false;
    this.sessionStartDate = new Date();
    this.scheduleNextBuzz();
    this.emit();
  }

  stop() {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
    this._state.isRunning = false;
    this.sessionStartDate = null;
    this.emit();
  }

  resumeFromBackground() {
    if (!this._state.isRunning || !this.sessionStartDate || !this.config) return;

    const now = new Date();
    const wallElapsed = Math.floor((now.getTime() - this.sessionStartDate.getTime()) / 1000);
    const missedSeconds = wallElapsed - this._state.elapsedSeconds;

    if (missedSeconds > 1) {
      this._state.elapsedSeconds = wallElapsed;

      while (this._state.elapsedSeconds >= this.nextBuzzAt && !this.checkComplete()) {
        this._state.reminderCount++;
        this.config.onInterval();
        if (this.checkComplete()) {
          this.config.onComplete?.();
          this.stop();
          return;
        }
        this.scheduleNextBuzz();
      }

      this._state.secondsUntilNextBuzz = Math.max(0, this.nextBuzzAt - this._state.elapsedSeconds);
    }

    if (!this.timer) {
      this.startTicking();
    }
    this.emit();
  }

  private startTicking() {
    if (this.timer) clearInterval(this.timer);
    this.timer = setInterval(() => {
      if (!this._state.isRunning || !this.config) return;

      this._state.elapsedSeconds++;
      this._state.secondsUntilNextBuzz = Math.max(0, this.nextBuzzAt - this._state.elapsedSeconds);

      if (this._state.elapsedSeconds >= this.nextBuzzAt) {
        this._state.reminderCount++;
        this.config.onInterval();

        if (this.checkComplete()) {
          this._state.isComplete = true;
          this.config.onComplete?.();
          this.stop();
        } else {
          this.scheduleNextBuzz();
        }
      }

      this.emit();
    }, 1000);
  }

  private scheduleNextBuzz() {
    if (!this.config) return;
    let next = this.config.intervalSeconds;
    if (this.config.varianceSeconds > 0) {
      const range = this.config.varianceSeconds * 2 + 1;
      next += Math.floor(Math.random() * range) - this.config.varianceSeconds;
      next = Math.max(5, next);
    }
    this.nextBuzzAt = this._state.elapsedSeconds + next;
    this._state.secondsUntilNextBuzz = next;
  }

  private checkComplete(): boolean {
    if (!this.config) return false;
    switch (this.config.endCondition.type) {
      case 'unlimited':
        return false;
      case 'afterCount':
        return this._state.reminderCount >= this.config.endCondition.count;
      case 'afterDuration':
        return this._state.elapsedSeconds >= this.config.endCondition.seconds;
    }
  }

  private emit() {
    this.onStateChange({ ...this._state });
  }
}

File: HabitCoach/Shared/Services/TimerService.swift
import Foundation
import os.log
#if os(iOS)
import AVFoundation
#endif

private let logger = Logger(subsystem: "com.ctuckersolutions.habitcoach", category: "TimerService")

enum SessionEndCondition: Codable, Hashable {
    case unlimited
    case afterCount(Int)
    case afterDuration(Int) // seconds
}

@MainActor @Observable
class TimerService {
    // State
    var isRunning: Bool = false
    var elapsedSeconds: Int = 0
    var reminderCount: Int = 0
    var secondsUntilNextBuzz: Int = 0

    // Configuration
    var intervalSeconds: Int = 60
    var varianceSeconds: Int = 0
    var endCondition: SessionEndCondition = .unlimited

    private var timer: Timer?
    private var nextBuzzAt: Int = 0
    private var onInterval: (() -> Void)?
    private var onComplete: (() -> Void)?

    // Background persistence
    private var sessionStartDate: Date?
    private var pausedAt: Date?

    #if os(iOS)
    /// Silent audio player that keeps the app alive in background
    private var backgroundPlayer: AVAudioPlayer?
    #endif

    var isComplete: Bool {
        switch endCondition {
        case .unlimited:
            return false
        case .afterCount(let count):
            return reminderCount >= count
        case .afterDuration(let duration):
            return elapsedSeconds >= duration
        }
    }

    var totalCycles: Int? {
        switch endCondition {
        case .unlimited: return nil
        case .afterCount(let count): return count
        case .afterDuration(let duration):
            guard intervalSeconds > 0 else { return nil }
            return duration / intervalSeconds
        }
    }

    func start(
        intervalSeconds: Int,
        varianceSeconds: Int = 0,
        endCondition: SessionEndCondition = .unlimited,
        onInterval: @escaping () -> Void,
        onComplete: (() -> Void)? = nil
    ) {
        guard intervalSeconds > 0 else {
            logger.error("Cannot start timer with interval <= 0")
            return
        }
        self.intervalSeconds = intervalSeconds
        self.varianceSeconds = varianceSeconds
        self.endCondition = endCondition
        self.onInterval = onInterval
        self.onComplete = onComplete
        self.elapsedSeconds = 0
        self.reminderCount = 0
        self.isRunning = true
        self.sessionStartDate = Date()
        self.pausedAt = nil

        #if os(iOS)
        startBackgroundAudio()
        #endif

        scheduleNextBuzz()
        startTicking()
    }

    /// Reconfigures the timer for a new step without stopping the underlying timer.
    /// Used by sequence playback to transition between steps seamlessly.
    func reconfigure(
        intervalSeconds: Int,
        varianceSeconds: Int = 0,
        endCondition: SessionEndCondition,
        onInterval: @escaping () -> Void,
        onComplete: (() -> Void)? = nil
    ) {
        self.intervalSeconds = intervalSeconds
        self.varianceSeconds = varianceSeconds
        self.endCondition = endCondition
        self.onInterval = onInterval
        self.onComplete = onComplete
        self.elapsedSeconds = 0
        self.reminderCount = 0
        self.sessionStartDate = Date()
        scheduleNextBuzz()
    }

    func stop() {
        timer?.invalidate()
        timer = nil
        isRunning = false
        sessionStartDate = nil
        pausedAt = nil
        #if os(iOS)
        stopBackgroundAudio()
        #endif
    }

    /// Called when app returns to foreground — recalculates elapsed time from wall clock
    func resumeFromBackground() {
        guard isRunning, let startDate = sessionStartDate else { return }

        let now = Date()
        let wallElapsed = Int(now.timeIntervalSince(startDate))
        let missedSeconds = wallElapsed - elapsedSeconds

        if missedSeconds > 1 {
            logger.info("Resuming from background, catching up \(missedSeconds)s")

            // Fast-forward elapsed time
            elapsedSeconds = wallElapsed

            // Count any missed buzzes
            while elapsedSeconds >= nextBuzzAt && !isComplete {
                reminderCount += 1
                onInterval?()
                if isComplete {
                    onComplete?()
                    stop()
                    return
                }
                scheduleNextBuzz()
            }

            secondsUntilNextBuzz = max(0, nextBuzzAt - elapsedSeconds)
        }

        // Restart the tick timer if it was invalidated
        if timer == nil {
            startTicking()
        }
    }

    private func startTicking() {
        timer?.invalidate()
        timer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            MainActor.assumeIsolated {
                guard let self, self.isRunning else { return }
                self.elapsedSeconds += 1
                self.secondsUntilNextBuzz = max(0, self.nextBuzzAt - self.elapsedSeconds)

                if self.elapsedSeconds >= self.nextBuzzAt {
                    self.reminderCount += 1
                    self.onInterval?()

                    if self.isComplete {
                        self.onComplete?()
                        self.stop()
                    } else {
                        self.scheduleNextBuzz()
                    }
                }
            }
        }
    }

    private func scheduleNextBuzz() {
        var next = intervalSeconds
        if varianceSeconds > 0 {
            let range = -varianceSeconds...varianceSeconds
            next += Int.random(in: range)
            next = max(5, next) // minimum 5 seconds between buzzes
        }
        nextBuzzAt = elapsedSeconds + next
        secondsUntilNextBuzz = next
    }

    // MARK: - Background Audio (iOS)
    // Plays a silent audio loop so iOS keeps the app alive in background.
    // This allows the timer and haptics to continue when the user switches apps.

    #if os(iOS)
    private func startBackgroundAudio() {
        do {
            let session = AVAudioSession.sharedInstance()
            try session.setCategory(.playback, mode: .default, options: [.mixWithOthers])
            try session.setActive(true)
        } catch {
            logger.error("Failed to configure audio session: \(error.localizedDescription)")
        }

        // Use cached silent WAV data (1 second of silence, 16-bit mono 44100 Hz)
        guard let silentData = Self.silentWAVData else {
            logger.error("Failed to generate silent audio data")
            return
        }

        do {
            let player = try AVAudioPlayer(data: silentData)
            player.numberOfLoops = -1 // loop forever
            player.volume = 0.0
            player.play()
            backgroundPlayer = player
            logger.info("Background audio started")
        } catch {
            logger.error("Failed to start background audio: \(error.localizedDescription)")
        }
    }

    private func stopBackgroundAudio() {
        backgroundPlayer?.stop()
        backgroundPlayer = nil
        try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
        logger.info("Background audio stopped")
    }

    /// Cached silent WAV data to avoid regenerating every session start.
    private static let silentWAVData: Data? = generateSilentWAV(durationSeconds: 1.0)

    /// Generates a minimal WAV file with silence, entirely in memory.
    private static func generateSilentWAV(durationSeconds: Double) -> Data? {
        let sampleRate: UInt32 = 44100
        let bitsPerSample: UInt16 = 16
        let channels: UInt16 = 1
        let numSamples = UInt32(Double(sampleRate) * durationSeconds)
        let dataSize = numSamples * UInt32(bitsPerSample / 8) * UInt32(channels)
        let fileSize = 36 + dataSize

        var data = Data()
        // RIFF header
        data.append(contentsOf: "RIFF".utf8)
        data.append(contentsOf: withUnsafeBytes(of: fileSize.littleEndian) { Array($0) })
        data.append(contentsOf: "WAVE".utf8)
        // fmt chunk
        data.append(contentsOf: "fmt ".utf8)
        data.append(contentsOf: withUnsafeBytes(of: UInt32(16).littleEndian) { Array($0) })
        data.append(contentsOf: withUnsafeBytes(of: UInt16(1).littleEndian) { Array($0) }) // PCM
        data.append(contentsOf: withUnsafeBytes(of: channels.littleEndian) { Array($0) })
        data.append(contentsOf: withUnsafeBytes(of: sampleRate.littleEndian) { Array($0) })
        let byteRate = sampleRate * UInt32(channels) * UInt32(bitsPerSample / 8)
        data.append(contentsOf: withUnsafeBytes(of: byteRate.littleEndian) { Array($0) })
        let blockAlign = channels * (bitsPerSample / 8)
        data.append(contentsOf: withUnsafeBytes(of: blockAlign.littleEndian) { Array($0) })
        data.append(contentsOf: withUnsafeBytes(of: bitsPerSample.littleEndian) { Array($0) })
        // data chunk
        data.append(contentsOf: "data".utf8)
        data.append(contentsOf: withUnsafeBytes(of: dataSize.littleEndian) { Array($0) })
        data.append(Data(count: Int(dataSize))) // silence (all zeros)
        return data
    }
    #endif
}
